src/runtime_src/core/common/xrt_xclbin.cpp

```diff
@@ -62,15 +67,342 @@ int
+}
+
+static void
+free_xclbin(xrtXclbinHandle handle)
+{
+  if (xclbins.erase(handle) == 0)
+    throw xrt_core::error(-EINVAL, "No such xclbin handle");
+}
+
+// Utility function for device class to verify that the C xclbin handle is valid
+// Needed when the C API for device tries to load an xclbin using C pointer to xclbin
+namespace xrt {
+namespace xclbin_int {
+bool
+is_valid(xrtXclbinHandle handle)
+{
+  bool rtn = (xclbins.find(handle) != xclbins.end());
+  return rtn;
+}
+}
+};
+
+xrtXclbinHandle
+xrtXclbinAllocFilename(const char *filename)
+{
+  try {
+    auto xclbin = std::make_shared<xrt::xclbin_impl>(filename);
+    auto handle = xclbin.get();
+    xclbins.emplace(std::make_pair(handle, std::move(xclbin)));
+    return handle;
+  }
+  catch (const xrt_core::error &ex) {
+    xrt_core::send_exception_message(ex.what());
+    errno = ex.get();
+  }
+  catch (const std::exception &ex) {
+    send_exception_message(ex.what());
+    errno = 0;
+  }
+  return nullptr;
+}
+
+xrtXclbinHandle
+xrtXclbinAllocRawData(const char *data, const int size)
+{
+  try {
+    char *data_c = const_cast<char *>(data);
+    std::vector<char> raw_data(data_c, data_c + size);
+    auto xclbin = std::make_shared<xrt::xclbin_impl>(raw_data);
+    auto handle = xclbin.get();
+    xclbins.emplace(std::make_pair(handle, std::move(xclbin)));
+    return handle;
+  }
+  catch (const xrt_core::error &ex) {
     xrt_core::send_exception_message(ex.what());
+    errno = ex.get();
+  }
+  catch (const std::exception &ex) {
+    send_exception_message(ex.what());
+    errno = 0;
+  }
+  return nullptr;
+}
+
+int
+xrtXclbinFreeHandle(xrtXclbinHandle handle)
+{
+  try {
+    free_xclbin(handle);
+    return 0;
+  }
+  catch (const xrt_core::error &ex) {
+    xrt_core::send_exception_message(ex.what());
+    return ex.get();
+  }
+  catch (const std::exception &ex) {
+    send_exception_message(ex.what());
     return -1;
   }
 }
+
+int
+xrtXclbinGetCUNames(xrtXclbinHandle handle, char **names, int *numNames)
+{
+  try {
+    auto xclbin = get_xclbin(handle);
+    const std::vector<std::string> cuNames = xclbin->get_cu_names();
+    // populate numNames if memory is allocated
+    if (numNames != nullptr)
+        *numNames = cuNames.size();
+    // populate names if memory is allocated
+    if (names != nullptr) {
+      auto index = 0;
+      for (auto &&name: cuNames) {
+        std::strcpy(names[index++], name.c_str());
+      }
+    }
+    return 0;
+  }
+  catch (const xrt_core::error &ex) {
+    xrt_core::send_exception_message(ex.what());
+    return (errno = ex.get());
+  }
+  catch (const std::exception &ex) {
+    send_exception_message(ex.what());
+    return (errno = 0);
+  }
+}
+
+int
+xrtXclbinGetXSAName(xrtXclbinHandle handle, char *name, int size, int *ret_size)
+{
+  try {
+    auto xclbin = get_xclbin(handle);
+    const std::string &xsaName = xclbin->get_xsa_name();
+    // populate ret_size if memory is allocated
+    if (ret_size != nullptr)
```

Comment:
Same as previous for the 2 if

Reply:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es87-dont-add-redundant--or--to-conditions

Reply:
replaced all != nullptr with ""